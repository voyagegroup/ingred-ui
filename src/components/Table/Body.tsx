import * as React from "react";

export type Props = React.ComponentPropsWithoutRef<"tbody">;

export const Body = React.forwardRef<HTMLTableSectionElement, Props>(
  function Body({ children, ...rest }, ref) {
    return (
      <tbody ref={ref} {...rest}>
        {children}
      </tbody>
    );
  },
);
