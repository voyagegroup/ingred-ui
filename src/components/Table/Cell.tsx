import * as React from "react";
import styled from "styled-components";
import Typography from "../Typography";

const Component = styled.td<{ width: string }>`
  width: ${({ width }) => width};
  padding: ${({ theme }) => theme.spacing * 2}px;
`;

export type TableCellProps =
  React.TdHTMLAttributes<HTMLTableDataCellElement> & {
    width?: string;
    children?: React.ReactNode;
  };

export const Cell = React.forwardRef<HTMLTableCellElement, TableCellProps>(
  function Cell({ width = "auto", children, ...rest }, ref) {
    return (
      <Component ref={ref} width={width} {...rest}>
        <Typography component="div">{children}</Typography>
      </Component>
    );
  },
);
