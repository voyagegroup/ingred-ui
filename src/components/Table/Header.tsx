import * as React from "react";

export type Props = {
  children: React.ReactNode;
};

export const Header = React.forwardRef<HTMLTableSectionElement, Props>(
  ({ children }, ref) => <thead ref={ref}>{children}</thead>,
);
