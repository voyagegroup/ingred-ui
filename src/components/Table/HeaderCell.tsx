import React from "react";
import styled from "styled-components";
import Typography from "../Typography";

const Component = styled.th<{ width: string }>`
  width: ${({ width }) => width};
  padding: ${({ theme }) => theme.spacing * 2}px;
  background-color: ${({ theme }) => theme.palette.gray.highlight};
  border: 1px solid ${({ theme }) => theme.palette.divider};
`;

export type TableHeaderCellProps =
  React.ThHTMLAttributes<HTMLTableHeaderCellElement> & {
    width?: string;
    children?: React.ReactNode;
  };

export const HeaderCell = React.forwardRef<
  HTMLTableCellElement,
  TableHeaderCellProps
>(function HeaderCell({ width = "auto", children, ...rest }, ref) {
  return (
    <Component ref={ref} width={width} {...rest}>
      <Typography component="div" weight="bold" size="md" lineHeight="normal">
        {children}
      </Typography>
    </Component>
  );
});
