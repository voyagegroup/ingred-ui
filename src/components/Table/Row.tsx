import * as React from "react";
import styled from "styled-components";

export type VerticalSpacing = "small" | "medium" | "large";

const verticalSpacingMap: { [key in VerticalSpacing]: string } = {
  small: "8px",
  medium: "12px",
  large: "16px",
};

type RowProps = {
  verticalSpacing: VerticalSpacing;
};

const Component = styled.tr<RowProps>`
  border: 1px solid ${({ theme }) => theme.palette.divider};

  & > td,
  th {
    padding-top: ${({ verticalSpacing }) =>
      verticalSpacingMap[verticalSpacing]};
    padding-bottom: ${({ verticalSpacing }) =>
      verticalSpacingMap[verticalSpacing]};
  }
`;

export type Props = React.ComponentPropsWithoutRef<"tr"> & Partial<RowProps>;

export const Row = React.forwardRef<HTMLTableRowElement, Props>(function Row(
  { children, verticalSpacing = "medium", ...rest },
  ref,
) {
  return (
    <Component ref={ref} verticalSpacing={verticalSpacing} {...rest}>
      {children}
    </Component>
  );
});
