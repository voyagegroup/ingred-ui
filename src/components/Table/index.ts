export { default } from "./Table";
export type { TableProps } from "./Table";
export type { TableCellProps } from "./Cell";
export type { TableHeaderCellProps } from "./HeaderCell";
