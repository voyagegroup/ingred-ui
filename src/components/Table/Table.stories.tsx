import React from "react";
import Badge from "../Badge";
import Flex from "../Flex";
import Typography from "../Typography";
import Table from "./";

export default {
  title: "Components/Data Display/Table",
  component: Table,
  subcomponents: {
    Header: Table.Header,
    Body: Table.Body,
    Row: Table.Row,
    Cell: Table.Cell,
    HeaderCell: Table.HeaderCell,
  },
};

export const Example = () => (
  <Table>
    <Table.Body>
      <Table.Row>
        <Table.HeaderCell>normal</Table.HeaderCell>
        <Table.HeaderCell>table</Table.HeaderCell>
      </Table.Row>
      <Table.Row>
        <Table.Cell>row</Table.Cell>
        <Table.Cell>row</Table.Cell>
      </Table.Row>
    </Table.Body>
  </Table>
);

export const ChangeVerticalSpacing = () => (
  <Table>
    <Table.Body>
      <Table.Row verticalSpacing="small">
        <Table.HeaderCell>
          <Flex
            display="flex"
            justifyContent="space-between"
            alignItems="center"
          >
            <Typography component="div" weight="bold" size="md">
              Title
            </Typography>
            <Badge color="danger" fontWeight="bold">
              Required
            </Badge>
          </Flex>
        </Table.HeaderCell>
        <Table.Cell>Contents</Table.Cell>
      </Table.Row>
      <Table.Row verticalSpacing="small">
        <Table.HeaderCell>Title</Table.HeaderCell>
        <Table.Cell>Contents</Table.Cell>
      </Table.Row>
    </Table.Body>
  </Table>
);

export const ChangeWidth = () => (
  <Table>
    <Table.Body>
      <Table.Row>
        <Table.HeaderCell width="100px">Title</Table.HeaderCell>
        <Table.Cell>Contents</Table.Cell>
      </Table.Row>
      <Table.Row>
        <Table.HeaderCell width="100px">Title</Table.HeaderCell>
        <Table.Cell>Contents</Table.Cell>
      </Table.Row>
    </Table.Body>
  </Table>
);

export const WithRowSpanAndColSpan = () => (
  <Table>
    <Table.Body>
      <Table.Row>
        <Table.HeaderCell>Title</Table.HeaderCell>
        <Table.Cell colSpan={2}>Contents</Table.Cell>
      </Table.Row>
      <Table.Row>
        <Table.HeaderCell>Title</Table.HeaderCell>
        <Table.Cell colSpan={2}>Contents</Table.Cell>
      </Table.Row>
      <Table.Row>
        <Table.HeaderCell rowSpan={2}>Title</Table.HeaderCell>
        <Table.HeaderCell>SubTitle</Table.HeaderCell>
        <Table.Cell>Contents</Table.Cell>
      </Table.Row>
      <Table.Row>
        <Table.HeaderCell>SubTitle</Table.HeaderCell>
        <Table.Cell>Contents</Table.Cell>
      </Table.Row>
    </Table.Body>
  </Table>
);
