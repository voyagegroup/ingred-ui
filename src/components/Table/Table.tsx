import * as React from "react";
import styled from "styled-components";
import { Body } from "./Body";
import { Cell } from "./Cell";
import { Header } from "./Header";
import { Row } from "./Row";
import { HeaderCell } from "./HeaderCell";

type ExportedComponentType = {
  (props: React.ComponentPropsWithRef<"table">): JSX.Element;
  Header: typeof Header;
  Body: typeof Body;
  Row: typeof Row;
  Cell: typeof Cell;
  HeaderCell: typeof HeaderCell;
};

const Container = styled.table`
  width: 100%;
  border-collapse: collapse;
  table-layout: auto;
  background-color: ${({ theme }) => theme.palette.background.default};
`;

export type TableProps = React.ComponentPropsWithoutRef<"table">;

const Table = React.forwardRef<HTMLTableElement, TableProps>(function Table(
  { children, ...rest },
  ref,
) {
  return (
    <Container ref={ref} {...rest}>
      {children}
    </Container>
  );
});

const ExportedComponent = Table as any;

ExportedComponent.Header = Header;
ExportedComponent.Body = Body;
ExportedComponent.Row = Row;
ExportedComponent.Cell = Cell;
ExportedComponent.HeaderCell = HeaderCell;

export default ExportedComponent as ExportedComponentType;
