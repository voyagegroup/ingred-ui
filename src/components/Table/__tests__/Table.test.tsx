import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";
import Table from "../";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("Table component testing", () => {
  afterEach(cleanup);

  test("Table", () => {
    const { asFragment } = renderWithThemeProvider(
      <Table>
        <Table.Body>
          <Table.Row>
            <Table.HeaderCell width="177px">Title</Table.HeaderCell>
            <Table.Cell>Content</Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.HeaderCell width="177px">Title</Table.HeaderCell>
            <Table.Cell>Content</Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.HeaderCell width="177px">Title</Table.HeaderCell>
            <Table.Cell>Content</Table.Cell>
          </Table.Row>
        </Table.Body>
      </Table>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
