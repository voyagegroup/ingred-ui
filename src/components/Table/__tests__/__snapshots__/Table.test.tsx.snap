// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component testing Table 1`] = `
<DocumentFragment>
  <table
    class="sc-eCImPb hTNeJW"
  >
    <tbody>
      <tr
        class="sc-dkPtRN gdWhAf"
      >
        <th
          class="sc-hKwDye drDPGO"
          width="177px"
        >
          <div
            class="sc-bdvvtL enNbws"
            color="#041C33"
            font-size="14px"
          >
            Title
          </div>
        </th>
        <td
          class="sc-gsDKAQ chzLmV"
          width="auto"
        >
          <div
            class="sc-bdvvtL euPFjW"
            color="#041C33"
            font-size="14px"
          >
            Content
          </div>
        </td>
      </tr>
      <tr
        class="sc-dkPtRN gdWhAf"
      >
        <th
          class="sc-hKwDye drDPGO"
          width="177px"
        >
          <div
            class="sc-bdvvtL enNbws"
            color="#041C33"
            font-size="14px"
          >
            Title
          </div>
        </th>
        <td
          class="sc-gsDKAQ chzLmV"
          width="auto"
        >
          <div
            class="sc-bdvvtL euPFjW"
            color="#041C33"
            font-size="14px"
          >
            Content
          </div>
        </td>
      </tr>
      <tr
        class="sc-dkPtRN gdWhAf"
      >
        <th
          class="sc-hKwDye drDPGO"
          width="177px"
        >
          <div
            class="sc-bdvvtL enNbws"
            color="#041C33"
            font-size="14px"
          >
            Title
          </div>
        </th>
        <td
          class="sc-gsDKAQ chzLmV"
          width="auto"
        >
          <div
            class="sc-bdvvtL euPFjW"
            color="#041C33"
            font-size="14px"
          >
            Content
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
